pkg/build/ignore/docker/dockign.go

```diff
@@ -0,0 +1,116 @@
+package docker
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/openshift/source-to-image/pkg/api"
+)
+
+type DOCKIGN struct {
+	Config *api.Config
+}
+
+func New(req *api.Config) *DOCKIGN {
+	d := &DOCKIGN{
+		Config: req,
+	}
+	return d
+}
+
+func (b *DOCKIGN) Ignore(config *api.Config) {
+	path := filepath.Join(config.WorkingDir, "upload", "src", api.DefaultIgnoreFile)
+	file, err := os.Open(path)
+	if err != nil {
+		glog.V(4).Infof("Ignore processing, problem opening %s because of %v\n", path, err)
+		return
+	}
+	defer file.Close()
+
+	// so, to duplicate the .dockerignore capabilities (https://docs.docker.com/reference/builder/#dockerignore-file)
+	// we have a flow that follows:
+	/*
+
+		0) First note, .dockerignore rules are NOT recursive (unlike .gitignore) .. you have to list subdir explicitly
+		1) Read in the exclusion patterns
+		2) Skip over comments (noted by #)
+		3) note overrides (via exclamation sign i.e. !) and reinstate files (don't remove) as needed
+		4) leverage Glob matching to build list, as .dockeringore is documented as following filepath.Match / filepath.Glob
+		5) del files
+
+	*/
+
+	filesToDel := make(map[string]string)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		filespec := scanner.Text()
+		glog.V(4).Infof(".s2iignore lists a file spec of %s \n", filespec)
+
+		if strings.HasPrefix(filespec, "#") {
+			continue
+		}
+		if strings.HasPrefix(filespec, "!") {
+			//remove any existing files to del that the override covers
+			// and patterns later on that undo this take precedence
+
+			// first, remove ! ... note, replace ! with */ did not have
+			// exected effect with filepath.Match
+			filespec = strings.Replace(filespec, "!", "", 1)
+
+			// iterate through and determine ones to leave in
+			dontDel := make([]string, 0)
+			for candidate, _ := range filesToDel {
+				// strip base for file name to file name compare
+				// NOTE, it would appear from git/docker ignore doc that you don't
+				// do something like !*/*/file.ext
```

Comment:
so ! arguments only apply to flie basenames, you can't specify a full directory?

eg i can't specify "!foo/bar/myfile"?

and if i specify "!myfile" it will not ignore (will keep) any file named "myfile" in any path?